File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AutoTPA.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.network.play.server.SPacketChat;

import java.util.function.Predicate;

/*
 * By Katatje 8 Dec 2019
 */

@Module.Info(name = "AutoTPA", description = "Auto Accepts/Declines TPA requests", category = Module.Category.MISC)
public class AutoTPA extends Module {

    private Setting<mode> mod = register(Settings.e("Response", mode.DENY));

    @EventHandler
    public Listener<PacketEvent.Receive> receiveListener;

    public AutoTPA() {
        this.receiveListener = new Listener<PacketEvent.Receive>(event -> {
            SPacketChat packet;
            if (event.getPacket() instanceof SPacketChat && (packet = (SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains(" has requested to teleport to you.")) {
                switch (mod.getValue()) {
                    case ACCEPT:
                        Wrapper.getPlayer().sendChatMessage("/tpaccept");
                        break;
                    case DENY:
                        Wrapper.getPlayer().sendChatMessage("/tpdeny");
                        break;
                }
            }
        }, (Predicate<PacketEvent.Receive>[]) new Predicate[0]);
    }

    public enum mode {
        ACCEPT, DENY
    }

}

File: src/main/kotlin/org/kamiblue/client/module/modules/chat/AutoTPA.kt
package org.kamiblue.client.module.modules.chat

import net.minecraft.network.play.server.SPacketChat
import org.kamiblue.client.event.events.PacketEvent
import org.kamiblue.client.manager.managers.FriendManager
import org.kamiblue.client.module.Category
import org.kamiblue.client.module.Module
import org.kamiblue.client.util.text.MessageDetection
import org.kamiblue.client.util.text.MessageSendHelper.sendServerMessage
import org.kamiblue.event.listener.listener

internal object AutoTPA : Module(
    name = "AutoTPA",
    description = "Automatically accept or decline /TPAs",
    category = Category.CHAT
) {
    private val friends = setting("Always Accept Friends", true)
    private val mode = setting("Response", Mode.DENY)

    private enum class Mode {
        ACCEPT, DENY
    }

    init {
        listener<PacketEvent.Receive> {
            if (it.packet !is SPacketChat || MessageDetection.Other.TPA_REQUEST detectNot it.packet.chatComponent.unformattedText) return@listener

            /* I tested that getting the first word is compatible with chat timestamp, and it as, as this is Receive and chat timestamp is after Receive */
            val name = it.packet.chatComponent.unformattedText.split(" ")[0]

            when (mode.value) {
                Mode.ACCEPT -> sendServerMessage("/tpaccept $name")
                Mode.DENY -> {
                    if (friends.value && FriendManager.isFriend(name)) {
                        sendServerMessage("/tpaccept $name")
                    } else {
                        sendServerMessage("/tpdeny $name")
                    }
                }
            }
        }
    }
}